### PR TITLE
Adding RefineBlockGenerator

### LINKED ISSUE #18686: Enable mesh generators to refine mesh per blocks
## Reason
Uniform refinement of the whole mesh is quite expensive for a lot of simulations. For example, a coupled conduction - fluid flow simulation of a reactor will have many elements in the outer part of the reactor, which will also be the costliest to refine, and where there is no flow. So assuming we are refining the mesh for fluid flow, this is problematic

Now while we already can use the Adaptivity system with a uniform marker per block to do this, it's a little complicated & convoluted, especially to doing multiple different levels of refinements on different blocks.

A mesh generator fits nicer in the workflow, works better with outputs. It ll also be more flexible, allowing refining at various stages in the mesh generation process. 

## Design
Add a RefineBlockGenerator

## Impact
Help for some scalability studies in a LDRD with ray tracing

### PR DESCRIPTION
**Pull request by socratesgorilla**
Adds a new MeshGenerator capable of refining individual blocks. Closes #18686

### CODE REVIEW

File: framework/include/meshgenerators/RefineBlockGenerator.h [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "MeshGenerator.h"
+
+/**
+ * MeshGenerator for removing blocks from the mesh
+ */
```
Comment (GiudGiud) [inline framework/include/meshgenerators/RefineBlockGenerator.h:16]: update that

File: framework/include/meshgenerators/RefineBlockGenerator.h [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+class RefineBlockGenerator : public MeshGenerator
+{
+public:
+  static InputParameters validParams();
+  virtual std::unique_ptr<MeshBase> recurs_refine(std::vector<subdomain_id_type> block_ids,
```
Comment (GiudGiud) [inline framework/include/meshgenerators/RefineBlockGenerator.h:21]: add docstring for this routine

recursive_refine is a better name

File: framework/include/meshgenerators/RefineBlockGenerator.h [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+                                                  std::unique_ptr<MeshBase> & mesh,
+                                                  std::vector<int> _refinement,
```
Comment (GiudGiud) [inline framework/include/meshgenerators/RefineBlockGenerator.h:23]: routine argument should not have _

File: framework/include/meshgenerators/RefineBlockGenerator.h (hunk 2 above)
Comment (GiudGiud) [reply framework/include/meshgenerators/RefineBlockGenerator.h:21]: this routine should be private too

File: framework/include/meshgenerators/RefineBlockGenerator.h [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+  RefineBlockGenerator(const InputParameters & parameters);
+
+protected:
+  virtual std::unique_ptr<MeshBase> generate() override;
+
+private:
+  std::unique_ptr<MeshBase> & _input;
+  std::vector<SubdomainName> _block;
+  std::vector<int> _refinement;
+  bool _enable_neighbors;
```
Comment (GiudGiud) [inline framework/include/meshgenerators/RefineBlockGenerator.h:36]: docstrings

Comment (GiudGiud) [inline framework/include/meshgenerators/RefineBlockGenerator.h:36]: enable_neighbor_refinement is more explicit

File: framework/src/meshgenerators/RefineBlockGenerator.C [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+#include "RefineBlockGenerator.h"
+#include "MooseMeshUtils.h"
+
+#include "libmesh/elem.h"
+#include "libmesh/mesh_refinement.h"
+#include "CastUniquePointer.h"
+
+registerMooseObject("MooseApp", RefineBlockGenerator);
+
+defineLegacyParams(RefineBlockGenerator);
```
Comment (GiudGiud) [inline framework/src/meshgenerators/RefineBlockGenerator.C:19]: no need for this

File: framework/src/meshgenerators/RefineBlockGenerator.C [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+
+InputParameters
+RefineBlockGenerator::validParams()
+{
+  InputParameters params = MeshGenerator::validParams();
+
+  params.addClassDescription("Mesh generator which removes elements from the specified subdomains");
```
Comment (GiudGiud) [inline framework/src/meshgenerators/RefineBlockGenerator.C:26]: update docstring

File: framework/src/meshgenerators/RefineBlockGenerator.C [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+  params.addRequiredParam<MeshGeneratorName>("input", "Input mesh to modify");
+  params.addRequiredParam<std::vector<SubdomainName>>("block", "The list of blocks to be modified");
```
Comment (GiudGiud) [inline framework/src/meshgenerators/RefineBlockGenerator.C:28]:
```suggestion
  params.addRequiredParam<std::vector<SubdomainName>>("block", "The list of blocks to be refined");
```

File: framework/src/meshgenerators/RefineBlockGenerator.C [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+
+RefineBlockGenerator::RefineBlockGenerator(const InputParameters & parameters)
+  : MeshGenerator(parameters),
+    _input(getMesh("input")),
+    _block(getParam<std::vector<SubdomainName>>("block")),
+    _refinement(getParam<std::vector<int>>("refinement")),
+    _enable_neighbors(getParam<bool>("enable_neighbor_refinement"))
+{
+}
+std::unique_ptr<MeshBase>
```
Comment (GiudGiud) [inline framework/src/meshgenerators/RefineBlockGenerator.C:48]:
```suggestion
}

std::unique_ptr<MeshBase>
```

File: framework/src/meshgenerators/RefineBlockGenerator.C [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+      else
+        paramError("block_id",
+                   "The block '",
+                   getParam<SubdomainID>("block_id"),
+                   "' was not found within the mesh");
+    }
+  std::unique_ptr<MeshBase> mesh = std::move(_input);
+  int max = *std::max_element(_refinement.begin(), _refinement.end());
+
+  return recurs_refine(block_ids, mesh, _refinement, max);
```
Comment (GiudGiud) [inline framework/src/meshgenerators/RefineBlockGenerator.C:77]: why pass a class attribute as an argument?

File: framework/include/meshgenerators/RefineBlockGenerator.h (hunk 3 above)
Comment (GiudGiud) [inline framework/include/meshgenerators/RefineBlockGenerator.h:23]: add constness qualifiers

File: test/tests/meshgenerators/refine_block_generator/tests [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+
+    [single_block]
+      type = 'Exodiff'
+      input = 'test_single.i'
+      exodiff = 'test_single_in.e'
+      cli_args = '--mesh-only'
+      recover = false
+      detail = 'in a mesh with a single first order element'
+
+    []
```
Comment (GiudGiud) [inline test/tests/meshgenerators/refine_block_generator/tests:16]: this is not what its doing

File: test/tests/meshgenerators/refine_block_generator/tests [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+    [multiple_block]
+      type = 'Exodiff'
+      input = 'test_multi.i'
+      exodiff = 'test_multi_in.e'
+      recover = false
+      cli_args = "--mesh-only"
+
+      detail = 'in a Cartesian 3D mesh.'
```
Comment (GiudGiud) [inline test/tests/meshgenerators/refine_block_generator/tests:24]: same, update

File: test/tests/meshgenerators/refine_block_generator/tests [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+    []
+    [neighbor_refinement]
+      type = 'Exodiff'
+      input = 'test_neighbor.i'
+      exodiff = 'test_neighbor_in.e'
+      cli_args = "--mesh-only"
+      recover = false
+      detail = 'with the neighbor refinement on'
```
Comment (GiudGiud) [inline test/tests/meshgenerators/refine_block_generator/tests:32]: describe more. 'the neighbor refinement on' feels like what we chose as a parameter. SQA is more about what the code does than what we named things

File: framework/doc/content/source/meshgenerators/RefineBlockGenerator.md [hunk 13]
```diff
@@ -0,0 +1,15 @@
+# ModifyNodeGenerator
```
Comment (GiudGiud) [inline framework/doc/content/source/meshgenerators/RefineBlockGenerator.md:1]: update

File: framework/doc/content/source/meshgenerators/RefineBlockGenerator.md [hunk 14]
```diff
@@ -0,0 +1,15 @@
...
+
+!syntax description /Mesh/RefineBlockGenerator
+
+## Overview
+
+This MeshGenerator object allows the user to refine one or more blocks in a mesh, with or without neighbor refinement.
```
Comment (GiudGiud) [inline framework/doc/content/source/meshgenerators/RefineBlockGenerator.md:7]:
```suggestion
This MeshGenerator object allows the user to refine one or more blocks in a mesh.
```
description later is good enough

**Review by GiudGiud — CHANGES_REQUESTED** [review]

File: framework/include/meshgenerators/RefineBlockGenerator.h [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+
+private:
+  /// Input mesh to refine
+  std::unique_ptr<MeshBase> & _input;
+
+  /// List of block(s) to refine
+  const std::vector<SubdomainName> _block;
+
+  /// The amount of times to refine each block, corresponding to their index in 'block'
+  const std::vector<int> _refinement;
```
Comment (andrsd) [inline framework/include/meshgenerators/RefineBlockGenerator.h:35]: Do you assume that refinement level would be negative? If not, this should be `unsigned int` then.

File: framework/src/meshgenerators/RefineBlockGenerator.C [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+    _block(getParam<std::vector<SubdomainName>>("block")),
+    _refinement(getParam<std::vector<int>>("refinement")),
+    _enable_neighbor_refinement(getParam<bool>("enable_neighbor_refinement"))
+{
+}
+
+std::unique_ptr<MeshBase>
+RefineBlockGenerator::generate()
+{
+
```
Comment (andrsd) [inline framework/src/meshgenerators/RefineBlockGenerator.C:50]: Remove this empty line.

File: framework/src/meshgenerators/RefineBlockGenerator.C [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+}
+
+std::unique_ptr<MeshBase>
+RefineBlockGenerator::recursive_refine(std::vector<subdomain_id_type> block_ids,
+                                       std::unique_ptr<MeshBase> & mesh,
+                                       std::vector<int> refinement,
+                                       int max,
+                                       int ref_step)
+{
+
```
Comment (andrsd) [inline framework/src/meshgenerators/RefineBlockGenerator.C:86]: Remove this empty line.

File: test/tests/meshgenerators/refine_block_generator/test_multi.i [hunk 18]
```diff
@@ -0,0 +1,31 @@
+[Mesh]
+  active='eg refine'
```
Comment (andrsd) [inline test/tests/meshgenerators/refine_block_generator/test_multi.i:2]: This should not be needed.

File: test/tests/meshgenerators/refine_block_generator/test_neighbor.i [hunk 19]
```diff
@@ -0,0 +1,30 @@
+[Mesh]
+  active='eg refine'
```
Comment (andrsd) [inline test/tests/meshgenerators/refine_block_generator/test_neighbor.i:2]: This should not be needed

File: test/tests/meshgenerators/refine_block_generator/test_single.i [hunk 20]
```diff
@@ -0,0 +1,31 @@
+[Mesh]
+  active='eg refine'
```
Comment (andrsd) [inline test/tests/meshgenerators/refine_block_generator/test_single.i:2]: This should not be needed either.

File: test/tests/meshgenerators/refine_block_generator/tests [hunk 21]
```diff
@@ -0,0 +1,33 @@
...
+    [neighbor_refinement]
+      type = 'Exodiff'
+      input = 'test_neighbor.i'
+      exodiff = 'test_neighbor_in.e'
+      cli_args = "--mesh-only"
+      recover = false
+      detail = 'multiple blocks within a multi-domain mesh, with neighbor refinement enabled'
+    []
+  []
+[]
```
Comment (andrsd) [inline test/tests/meshgenerators/refine_block_generator/tests:33]: At least `test_multi.i` and `test_neigbor.i` can be just one file and then you can pass the `enable_neighbor_refinement` parameter via `cli_args` like so:
```
cli_args = 'Mesh/refine/enable_neighbor_refinement=false'
```

You could take that one step further and even merge all 3 input files into one, but that would be optional

File: framework/include/meshgenerators/RefineBlockGenerator.h [hunk 22]
```diff
@@ -0,0 +1,46 @@
...
+
+  /// Toggles whether neighboring level one elements should be refined or not. Defaults to true.
+  const bool _enable_neighbor_refinement;
+
+  /// The actual function refining the blocks. This is done recursively in order to minimize the number of refinement iterations to as little as possible.
```
Comment (andrsd) [inline framework/include/meshgenerators/RefineBlockGenerator.h:40]: Methods use `/** */` comments and would need things like  `@return` and `@param`

**Review by andrsd — CHANGES_REQUESTED** [review]

File: test/tests/meshgenerators/refine_block_generator/tests (hunk 21 above)
Comment (GiudGiud) [reply test/tests/meshgenerators/refine_block_generator/tests:33]: he had it this way (merged) initially but in --mesh-only the file_base in output is being ignored.
We could if we turn this into a solve and not a mesh-only case. 
Please let us know what you prefer

Comment (GiudGiud) [reply test/tests/meshgenerators/refine_block_generator/tests:33]: ok seems you can pass the name of the output file to mesh only, like this
--mesh-only "hex_2d_err.e"'
@socratesgorilla let s do that

**Review by GiudGiud — APPROVED** [review]

**Review by andrsd — APPROVED** [review]

**Pull request merged**